salt/cloud/__init__.py

```diff
@@ -231,7 +231,10 @@ def _opts_defaults(self, **kwargs):
                          if a.get('provider', '')]
             if providers:
                 _providers = opts.get('providers', {})
-                for provider in _providers:
+
+                # keys() is required here because the list of providers is
+                # modified inside the loop
+                for provider in _providers.keys():
```

Comment:
Then `list()` on it, instead of `.keys()` or disable the lint error.
Actually, `list()`, since on python 3 `.keys()` is an iterator over the keys and not a list copy of the keys, so this issue will still happen in py3.